cmd/kubeadm/app/phases/addons/addons.go

```diff
@@ -69,14 +66,13 @@ func CreateEssentialAddons(cfg *kubeadmapi.MasterConfiguration, client *clientse
 		return fmt.Errorf("error when parsing kube-dns deployment template: %v", err)
 	}
 
-	// Get the DNS IP
-	dnsip, err := getDNSIP(cfg.Networking.ServiceSubnet)
+	dnsip, err := getDNSIP(client)
 	if err != nil {
 		return err
 	}
 
 	dnsServiceBytes, err := kubeadmutil.ParseTemplate(KubeDNSService, struct{ DNSIP string }{
-		DNSIP: dnsip.String(),
+		DNSIP: dnsip,
```

Comment:
See other comment about net.IP.  I see why you did it, but I would investigate changing DNSIP to type net.IP. For me, it make more sense to store an IP as a net.IP type than a string.